## Data Preprocessing in brief


In [133]:
# Importing all the required libraries
# One Hot Encoding is done using pandas itself so sklearn is not used here

import calendar
import pandas as pd
import numpy as np
from datetime import datetime

In [134]:
# Loading the dataset using pandas

dataFrame = pd.read_csv("lending-club-loans.csv")
dataFrame


72176    70868.0     2525.0   
42535                                            71623    70735.0     6500.0   
42536                                          1071795  1306957.0     5600.0   
42537                                            85961    85923.0     1200.0   
42538                                           427722   505512.0     5000.0   
42539                                           691253   881809.0    10000.0   
42540                                           714553   907984.0    16000.0   
42541                                           852787  1064904.0     8300.0   
42542                                           725123   920334.0     7000.0   
42543                                            74323    74301.0     6500.0   
42544                                           799847  1005087.0     3200.0   
42545                                           782452   985441.0    20000.0   
42546                                           670124   856744.0    20000.0   
42547                                           567541   730106.0    16000.0   
42548                                           505136   650770.0    20000.0   
42549                                           473367   598172.0     1750.0   
42550                                            70686    70681.0     5000.0   
42551  Total amount funded in policy code 1: 460296150        NaN        NaN   
42552          Total amount funded in policy code 2: 0        NaN        NaN   

       funded_amnt  funded_amnt_inv        term int_rate  installment grade  \
0           5000.0      4975.000000   36 months   10.65%       162.87     B   
1           2500.0      2500.000000   60 months   15.27%        59.83     C   
2           2400.0      2400.000000   36 months   15.96%        84.33     C   
3          10000.0     10000.000000   36 months   13.49%       339.31     C   
4           3000.0      3000.000000   60 months   12.69%        67.79     B   
5           5000.0      5000.000000   36 months    7.90%       156.46     A   
6           7000.0      7000.000000   60 months   15.96%       170.08     C   
7           3000.0      3000.000000   36 months   18.64%       109.43     E   
8           5600.0      5600.000000   60 months   21.28%       152.39     F   
9           5375.0      5350.000000   60 months   12.69%       121.45     B   
10          6500.0      6500.000000   60 months   14.65%       153.45     C   
11         12000.0     12000.000000   36 months   12.69%       402.54     B   
12          9000.0      9000.000000   36 months   13.49%       305.38     C   
13          3000.0      3000.000000   36 months    9.91%        96.68     B   
14         10000.0     10000.000000   36 months   10.65%       325.74     B   
15          1000.0      1000.000000   36 months   16.29%        35.31     D   
16         10000.0     10000.000000   36 months   15.27%       347.98     C   
17          3600.0      3600.000000   36 months    6.03%       109.57     A   
18          6000.0      6000.000000   36 months   11.71%       198.46     B   
19          9200.0      9200.000000   36 months    6.03%       280.01     A   
20         20250.0     19142.161080   60 months   15.27%       484.63     C   
21         21000.0     21000.000000   36 months   12.42%       701.73     B   
22         10000.0     10000.000000   36 months   11.71%       330.76     B   
23         10000.0     10000.000000   36 months   11.71%       330.76     B   
24          6000.0      6000.000000   36 months   11.71%       198.46     B   
25         15000.0     15000.000000   36 months    9.91%       483.38     B   
26         15000.0      8725.000000   36 months   14.27%       514.64     C   
27          5000.0      5000.000000   60 months   16.77%       123.65     D   
28          4000.0      4000.000000   36 months   11.71%       132.31     B   
29          8500.0      8500.000000   36 months   11.71%       281.15     B   
...            ...              ...         ...      ...          ...   ...   


In [135]:
# Overview of the dataset

print("Dataset Shape:",dataFrame.shape)
print("\nDataset info:")
print(dataFrame.info())

Dataset Shape: (42553, 115)

Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42553 entries, 0 to 42552
Columns: 115 entries, id to total_il_high_credit_limit
dtypes: float64(86), object(29)
memory usage: 37.3+ MB
None


We can see here that in the 115 columns, 86 are of float type and others are object types(like strings).
We need to change all the datatypes to the required ones by ourselves instead of letting pandas determine it in order to suit our convenience.
Before that, we can do other preprocessing steps.

In [136]:
# Check the presence of NULL values

dataFrame.isna()

# We are able to see a lot of True values => NULL values are there

e      False        False            False  False   
17     False      False      False        False            False  False   
18     False      False      False        False            False  False   
19     False      False      False        False            False  False   
20     False      False      False        False            False  False   
21     False      False      False        False            False  False   
22     False      False      False        False            False  False   
23     False      False      False        False            False  False   
24     False      False      False        False            False  False   
25     False      False      False        False            False  False   
26     False      False      False        False            False  False   
27     False      False      False        False            False  False   
28     False      False      False        False            False  False   
29     False      False      False        False            False  False   
...      ...        ...        ...          ...              ...    ...   
42523  False      False      False        False            False  False   
42524  False      False      False        False            False  False   
42525  False      False      False        False            False  False   
42526  False      False      False        False            False  False   
42527  False      False      False        False            False  False   
42528  False      False      False        False            False  False   
42529  False      False      False        False            False  False   
42530  False      False      False        False            False  False   
42531  False      False      False        False            False  False   
42532  False      False      False        False            False  False   
42533  False      False      False        False            False  False   
42534  False      False      False        False            False  False   
42535  False      False      False        False            False  False   
42536  False      False      False        False            False  False   
42537  False      False      False        False            False  False   
42538  False      False      False        False            False  False   
42539  False      False      False        False            False  False   
42540  False      False      False        False            False  False   
42541  False      False      False        False            False  False   
42542  False      False      False        False            False  False   
42543  False      False      False        False            False  False   
42544  False      False      False        False            False  False   
42545  False      False      False        False            False  False   
42546  False      False      False        False            False  False   
42547  False      False      False        False            False  False   
42548  False      False      False        False            False  False   
42549  False      False      False        False            False  False   
42550  False      False      False        False            False  False   
42551  False       True       True         True             True   True   
42552  False       True       True         True             True   True   

       int_rate  installment  grade  sub_grade  ...  num_tl_90g_dpd_24m  \
0         False        False  False      False  ...                True   
1         False        False  False      False  ...                True   
2         False        False  False      False  ...                True   
3         False        False  False      False  ...                True   
4         False        False  False      False  ...                True   
5         False        False  False      False  ...                True   
6         False        False  False      False  ...                True   
7         False        False  False      False

In [137]:
# Since we have a large dataset ~30MB, we can instead display their count

dataFrame.isna().sum()

id                                    0
member_id                             3
loan_amnt                             3
funded_amnt                           3
funded_amnt_inv                       3
term                                  3
int_rate                              3
installment                           3
grade                                 3
sub_grade                             3
emp_title                          2630
emp_length                         1115
home_ownership                        3
annual_inc                            7
verification_status                   3
issue_d                               3
loan_status                           3
pymnt_plan                            3
url                                   3
desc                              13298
purpose                               3
title                                16
zip_code                              3
addr_state                            3
dti                                   3


We can see that there are a lot of columns with the whole value in them as NULL.

In [138]:
# Removing whole columns where number of NULL values is more than half the number of rows

dataFrame = dataFrame.dropna(axis=1,thresh=dataFrame.shape[0]/2)
dataFrame.shape

(42553, 58)

After the dropping of columns operation, we have managed to reduce columns from 115 to 58.

In [139]:
dataFrame.isna().sum() > 10000

id                            False
member_id                     False
loan_amnt                     False
funded_amnt                   False
funded_amnt_inv               False
term                          False
int_rate                      False
installment                   False
grade                         False
sub_grade                     False
emp_title                     False
emp_length                    False
home_ownership                False
annual_inc                    False
verification_status           False
issue_d                       False
loan_status                   False
pymnt_plan                    False
url                           False
desc                           True
purpose                       False
title                         False
zip_code                      False
addr_state                    False
dti                           False
delinq_2yrs                   False
earliest_cr_line              False
fico_range_low              

In [140]:
# The desc column seems to have more than 10,000 rows as NULL
# Since it seems to have only description strings like the one below, we can drop it too

print(dataFrame.iloc[3]['desc'])
dataFrame = dataFrame.drop(labels=['desc'],axis=1)

  Borrower added on 12/21/11 > to pay for property tax (borrow from friend, need to pay back) & central A/C need to be replace. I'm very sorry to let my loan expired last time.<br>


In [141]:
dataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42553 entries, 0 to 42552
Data columns (total 57 columns):
id                            42553 non-null object
member_id                     42550 non-null float64
loan_amnt                     42550 non-null float64
funded_amnt                   42550 non-null float64
funded_amnt_inv               42550 non-null float64
term                          42550 non-null object
int_rate                      42550 non-null object
installment                   42550 non-null float64
grade                         42550 non-null object
sub_grade                     42550 non-null object
emp_title                     39923 non-null object
emp_length                    41438 non-null object
home_ownership                42550 non-null object
annual_inc                    42546 non-null float64
verification_status           42550 non-null object
issue_d                       42550 non-null object
loan_status                   42550 non-null object
p

In [142]:
dataFrame.nunique()
dataFrame['emp_length'].unique()

array(['10+ years', '< 1 year', '1 year', '3 years', '8 years', '9 years',
       '4 years', '5 years', '6 years', '2 years', '7 years', nan],
      dtype=object)

Thus, we can remove columns with same value for all rows.

In [143]:
dataFrame = dataFrame.drop( dataFrame.columns[np.where(dataFrame.nunique() == 1)[0]] ,axis=1)

In [144]:
# Next we should remove the columns with nearly same values throughout all rows

remove_arr = []

for i in dataFrame.columns.to_list():
    temp = dataFrame[i].value_counts()
    for j in temp:
        if j>30000:
            remove_arr.append(i)
# Here we can remove these columns because they are not related to the target variable => Loan Status
# Also because some columns like title have as many as 21,000 unique values => expensive to map them to integer values
remove_arr = remove_arr + ['id','member_id','zip_code','emp_title','url','title'] 

# Here we can remove grade since it is a redundant column of sub_grade
remove_arr = remove_arr + ['grade']

remove_arr


['term',
 'loan_status',
 'pymnt_plan',
 'delinq_2yrs',
 'pub_rec',
 'out_prncp',
 'out_prncp_inv',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'acc_now_delinq',
 'delinq_amnt',
 'pub_rec_bankruptcies',
 'tax_liens',
 'id',
 'member_id',
 'zip_code',
 'emp_title',
 'url',
 'title',
 'grade']

In [145]:
dataFrame = dataFrame.drop(labels=remove_arr,axis=1)

In [146]:
# dataFrame.isna().sum()
# Here since the same rows are responsilble for NULL values. we can remove them

print(np.where(dataFrame['loan_amnt'].isna()))
print(np.where(dataFrame['int_rate'].isna()))
print(np.where(dataFrame['funded_amnt_inv'].isna()))
print(np.where(dataFrame['funded_amnt'].isna()))

(array([39786, 42551, 42552], dtype=int64),)
(array([39786, 42551, 42552], dtype=int64),)
(array([39786, 42551, 42552], dtype=int64),)
(array([39786, 42551, 42552], dtype=int64),)


In [147]:
ind = np.where(dataFrame['loan_amnt'].isna())[0]

dataFrame.drop(dataFrame.index[ind],inplace=True)

dataFrame.isna().sum()


loan_amnt                  0
funded_amnt                0
funded_amnt_inv            0
int_rate                   0
installment                0
sub_grade                  0
emp_length              1112
home_ownership             0
annual_inc                 4
verification_status        0
issue_d                    0
purpose                    0
addr_state                 0
dti                        0
earliest_cr_line          31
fico_range_low             0
fico_range_high            0
inq_last_6mths            31
open_acc                  31
revol_bal                  0
revol_util                92
total_acc                 31
total_pymnt                0
total_pymnt_inv            0
total_rec_prncp            0
total_rec_int              0
last_pymnt_d              83
last_pymnt_amnt            0
last_credit_pull_d         4
last_fico_range_high       0
last_fico_range_low        0
dtype: int64

In [148]:
# Also, we can remove the rows with NULL dates

ind = np.where(dataFrame.earliest_cr_line.isna())[0]
dataFrame.drop(dataFrame.index[ind],inplace=True)

ind = np.where(dataFrame.last_pymnt_d.isna())[0]
dataFrame.drop(dataFrame.index[ind],inplace=True)

ind = np.where(dataFrame.last_credit_pull_d.isna())[0]
dataFrame.drop(dataFrame.index[ind],inplace=True)

dataFrame.isna().sum()

loan_amnt                  0
funded_amnt                0
funded_amnt_inv            0
int_rate                   0
installment                0
sub_grade                  0
emp_length              1107
home_ownership             0
annual_inc                 0
verification_status        0
issue_d                    0
purpose                    0
addr_state                 0
dti                        0
earliest_cr_line           0
fico_range_low             0
fico_range_high            0
inq_last_6mths             0
open_acc                   0
revol_bal                  0
revol_util                57
total_acc                  0
total_pymnt                0
total_pymnt_inv            0
total_rec_prncp            0
total_rec_int              0
last_pymnt_d               0
last_pymnt_amnt            0
last_credit_pull_d         0
last_fico_range_high       0
last_fico_range_low        0
dtype: int64

In [151]:
# Change % string to float

dataFrame['int_rate'] = pd.to_numeric(dataFrame['int_rate'].str.strip('%')).div(100)
dataFrame['int_rate'].head()

0    0.1065
1    0.1527
2    0.1596
3    0.1349
4    0.1269
Name: int_rate, dtype: float64

In [152]:
# Map sub_grade to numbers

print(dataFrame['sub_grade'].unique())

grades = { 'A1': 35, 'A2': 34,'A3': 33,'A4': 32,'A5': 31,
           'B1': 30, 'B2': 29,'B3': 28,'B4': 27,'B5': 26,
           'C1': 25, 'C2': 24,'C3': 23,'C4': 22,'C5': 21,
           'D1': 20, 'D2': 19,'D3': 18,'D4': 17,'D5': 16,
           'E1': 15, 'E2': 14,'E3': 13,'E4': 12,'E5': 11,
           'F1': 10, 'F2': 9,'F3': 8,'F4': 7,'F5': 6,
           'G1': 5, 'G2': 4,'G3': 3,'G4': 2,'G5': 1
  }

dataFrame['sub_grade'] = dataFrame['sub_grade'].map(grades)
dataFrame['sub_grade'].head()

['B2' 'C4' 'C5' 'C1' 'B5' 'A4' 'E1' 'F2' 'C3' 'B1' 'D1' 'A1' 'B3' 'B4'
 'D2' 'A3' 'A5' 'D5' 'A2' 'E4' 'D3' 'C2' 'D4' 'F3' 'E3' 'F4' 'F1' 'E5'
 'G4' 'E2' 'G3' 'G2' 'G1' 'F5' 'G5']


0    29
1    22
2    21
3    25
4    26
Name: sub_grade, dtype: int64

In [153]:
# Map experience in years to int

temp = { '10+ years': 10, '9 years': 9,'8 years': 8,'7 years': 7,'6 years': 6, '5 years':5, '4 years':4 , '3 years':3, '2 years':2, '1 year':1, '< 1 year':0  }

dataFrame['emp_length'] = dataFrame['emp_length'].map(temp)
dataFrame['emp_length'].head()

0    10.0
1     0.0
2    10.0
3    10.0
4     1.0
Name: emp_length, dtype: float64

In [154]:
# Get all columns with object datatype

dftypes = dataFrame.columns.groupby(dataFrame.dtypes)
dfobjects = dftypes[np.dtype('O')]
dfobjects

Index(['home_ownership', 'verification_status', 'issue_d', 'purpose',
       'addr_state', 'earliest_cr_line', 'last_pymnt_d', 'last_credit_pull_d'],
      dtype='object')

In [155]:
# Now for the rest with less than 50 unique values we can use One Hot Encoding since these don't have any numerical order

for i in dfobjects:
    if dataFrame[i].nunique() <= 50:
        print(i, dataFrame[i].unique())
        print()

home_ownership ['RENT' 'OWN' 'MORTGAGE' 'OTHER' 'NONE']

verification_status ['Verified' 'Source Verified' 'Not Verified']

purpose ['credit_card' 'car' 'small_business' 'other' 'wedding'
 'debt_consolidation' 'home_improvement' 'major_purchase' 'medical'
 'moving' 'vacation' 'house' 'renewable_energy' 'educational']

addr_state ['AZ' 'GA' 'IL' 'CA' 'OR' 'NC' 'TX' 'VA' 'MO' 'CT' 'UT' 'FL' 'PA' 'MN'
 'NY' 'NJ' 'KY' 'OH' 'SC' 'RI' 'LA' 'MA' 'WA' 'WI' 'AL' 'CO' 'KS' 'NV'
 'AK' 'MD' 'WV' 'VT' 'MI' 'DC' 'SD' 'NH' 'AR' 'NM' 'MT' 'HI' 'WY' 'OK'
 'DE' 'MS' 'TN' 'IA' 'NE' 'ID' 'IN' 'ME']



In [156]:
# One Hot Encoding for the above columns can be done by: pandas.get_dummies method
# Later we can drop and merge the encoded columns to original ones
# Please see dataset for better view

a = pd.get_dummies(dataFrame.home_ownership ,prefix='Home_Ownership')
b = pd.get_dummies(dataFrame.verification_status ,prefix='Verification_Status')
c = pd.get_dummies(dataFrame.purpose ,prefix='Purpose')
d = pd.get_dummies(dataFrame.addr_state ,prefix='State')

dataFrame.drop(labels=['home_ownership','verification_status','purpose','addr_state'],axis=1,inplace=True)

dataFrame = pd.concat([dataFrame,a,b,c,d],axis=1)
del a,b,c,d

In [157]:
# These functions are written to change a date string in the form 'Mmm-yy' to seconds

def year_change(row):
        if row>20:
            return row + 1900
        else:
            return row + 2000

# This function returns seconds from the month number and year
# Assumes that date is 1st of the month and all h,min,sec,msec are all 0
def change_to_seconds(row):
    # As I told, if year is < 1971 I reset it to 1971
    if row[1] < 1971:
        year = 1971
    else:
        year = row[1]
    return datetime(int(year),int(row[0]),1,0,0,0,0).timestamp()

# By convention, the seconds conversion is done from 1970 Jan 1st so, I am following it here too
def add_seconds_from_date_string(df,col_name,new_col_name):
    # Splits the date strings to Mmm and yy columns
    a = df[col_name].str.split('-',expand=True)
    # Changes the year from yy to yyyy, ie, 19 to 2019 and 75 to 1975
    a[1] = a[1].apply(pd.to_numeric)
    a[1] = a[1].apply(year_change)
    # This below is for changing Mmm to month number
    d = dict((v,k) for k,v in enumerate(calendar.month_abbr))
    a[0] = a[0].map(d)
    # Applies the function and creates the seconds column
    a['sec'] = a.apply(change_to_seconds,axis=1)
    # Drops the other extra columns
    df[new_col_name] = a.drop(labels=[0,1],axis=1)
    df.drop(labels=[col_name],axis=1,inplace=True)
    del a

In [158]:
# Change all the dates to seconds in int

print("Before change: \n")
print(dataFrame['earliest_cr_line'].head())
add_seconds_from_date_string(dataFrame,'issue_d','issue_seconds')
add_seconds_from_date_string(dataFrame,'earliest_cr_line','earliest_cr_line_seconds')
add_seconds_from_date_string(dataFrame,'last_pymnt_d','last_pymnt_seconds')
add_seconds_from_date_string(dataFrame,'last_credit_pull_d','last_credit_pull_seconds')
print("After change: \n")
print(dataFrame['earliest_cr_line_seconds'].head())

Before change: 

0    Jan-85
1    Apr-99
2    Nov-01
3    Feb-96
4    Jan-96
Name: earliest_cr_line, dtype: object
After change: 

0    4.733658e+08
1    9.229050e+08
2    1.004553e+09
3    8.231130e+08
4    8.204346e+08
Name: earliest_cr_line_seconds, dtype: float64


In [159]:
dataFrame.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 42433 entries, 0 to 42549
Data columns (total 99 columns):
loan_amnt                              42433 non-null float64
funded_amnt                            42433 non-null float64
funded_amnt_inv                        42433 non-null float64
int_rate                               42433 non-null float64
installment                            42433 non-null float64
sub_grade                              42433 non-null int64
emp_length                             41326 non-null float64
annual_inc                             42433 non-null float64
dti                                    42433 non-null float64
fico_range_low                         42433 non-null float64
fico_range_high                        42433 non-null float64
inq_last_6mths                         42433 non-null float64
open_acc                               42433 non-null float64
revol_bal                              42433 non-null float64
revol_util               

Now, all have been converted to either int or float as mentionned.
Finally, we can check for and replace NULL values with mean in the int columns.

In [160]:
dataFrame.columns[ np.where(dataFrame.isna().sum()>0)[0] ]

Index(['emp_length', 'revol_util'], dtype='object')

In [161]:
# Fill in the NULL values in the columns with their mean.

dataFrame = dataFrame.fillna(dataFrame.mean())

In [162]:
dataFrame.columns[ np.where(dataFrame.isna().sum()>0)[0] ]
# We can see that now all the NULL vlaues are gone successfully.

Index([], dtype='object')

In [163]:
# Check and count duplicate rows
# Don't understand why but this returns wrong values

print("No of duplicated rows: ",np.count_nonzero(dataFrame.duplicated()))
print("Duplicated rows: ")

arr = np.where(dataFrame.duplicated(keep='first'))[0]

dataFrame.iloc[arr]

No of duplicated rows:  13
Duplicated rows: 


loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  \
42528     5000.0       5000.0      5000.000000    0.0692       154.21   
42536     5600.0       5600.0      5600.000000    0.2128       152.39   
42538     5000.0       5000.0      4892.425415    0.1322       169.01   
42539    10000.0      10000.0      9850.000000    0.0692       308.41   
42540    16000.0      16000.0     16000.000000    0.1714       398.85   
42541     8300.0       8300.0      8300.000000    0.0849       261.98   
42542     7000.0       7000.0      7000.000000    0.0692       215.89   
42544     3200.0       3200.0      3200.000000    0.0699        98.80   
42545    20000.0      20000.0     19903.396040    0.1299       454.96   
42546    20000.0      20000.0     14999.996530    0.1751       502.56   
42547    16000.0      16000.0     15605.384280    0.0691       316.15   
42548    20000.0      20000.0     19500.000000    0.0988       644.21   
42549     1750.0       1750.0      1750.000000    0.1183        57.99   

       sub_grade  emp_length  annual_inc    dti  fico_range_low  ...  \
42528         33         9.0     48000.0  14.93           765.0  ...   
42536          9         4.0     40000.0   5.55           675.0  ...   
42538         24         1.0     80000.0  11.85           685.0  ...   
42539         33         0.0     87200.0  10.68           760.0  ...   
42540         13         8.0     83202.0   9.23           690.0  ...   
42541         31         5.0     50000.0   2.88           735.0  ...   
42542         33         4.0     24000.0   5.55           760.0  ...   
42544         33         3.0     83000.0  11.15           730.0  ...   
42545         25         5.0     74872.0  22.69           745.0  ...   
42546         12         3.0     64480.0  15.97           695.0  ...   
42547         31         2.0     47500.0  10.64           735.0  ...   
42548         30         0.0     55000.0   3.34           800.0  ...   
42549         28         3.0     40000.0  22.44           705.0  ...   

       State_VA  State_VT  State_WA  State_WI  State_WV  State_WY  \
42528         0         0         0         0         0         0   
42536         0         0         0         0         0         0   
42538         0         0         0         0         0         0   
42539         0         0         0         0         0         0   
42540         0         0         0         0         0         0   
42541         0         0         0         0         0         0   
42542         0         0         0         0         0         0   
42544         0         0         0         0         0         0   
42545         0         0         0         0         0         0   
42546         0         0         0         0         0         0   
42547         0         0         0         0         0         0   
42548         0         0         0         0         0         0   
42549         0         0         0         0         0         0   

       issue_seconds  earliest_cr_line_seconds  last_pymnt_seconds  \
42528   1.301596e+09              1.044038e+09        1.383244e+09   
42536   1.322678e+09              1.080758e+09        1.333219e+09   
42538   1.251743e+09              9.098586e+08        1.346438e+09   
42539   1.298918e+09              9.703386e+08        1.304188e+09   
42540   1.301596e+09              9.886554e+08        1.383244e+09   
42541   1.312137e+09              1.138732e+09        1.320086e+09   
42542   1.301596e+09              7.205562e+08        1.398883e+09   
42544   1.309459e+09              9.756090e+08        1.320086e+09   
42545   1.306867e+09              4.205754e+08        1.354300e+09   
42546   1.296499e+09              2.286810e+08        1.325356e+09   
42547   1.285871e+09              9.886554e+08        1.388515e+09   
42548   1.272652e+09              8.072154e+08        1.367347e+09   
42549   1.262284e+09              7.178778e+08        1.288550e+09   

       last_credit_pull_sec

In [164]:
# Saving to take a look at it

dataFrame.to_csv('Modified-Final.csv')

## This is not the end?!
